### PR TITLE
Call the stop() method on the server no matter how PHP shutsdown

### LINKED ISSUE #120: PHP fatal error in context file leads to node process stucking
I have such case: PHP fatal error in context (like attempt to click not existing button `$this->getSession()->getPage()->findButton('NotExisting')->click();`) leads to `Node` process stucking - repeated `Behat` start fails because `Node` can not reuse port which is already in use.

First start

```
# php ./composer/vendor/behat/behat/bin/behat

...[feature description skipped]...

PHP Fatal error:  Call to a member function click() on a non-object 
in /project/tests/behat/features/bootstrap/Search.php on line 17

Fatal error: Call to a member function click() on a non-object 
in /project/tests/behat/features/bootstrap/Search.php on line 17
```

Second start and further

```
# php ../../composer/vendor/behat/behat/bin/behat

...[feature description skipped]...

Server process has been terminated: (8) [
events.js:72
      throw er; // Unhandled 'error' event
            ^
Error: listen EADDRINUSE
  at errnoException (net.js:904:11)
  at Server._listen2 (net.js:1042:14)
  at listen (net.js:1064:10)
  at net.js:1146:9
  at Object._onImmediate (/usr/lib/node_modules/zombie/lib/zombie/dns_mask.js:69:16)
  at processImmediate [as _immediateCallback] (timers.js:345:15)
] (RuntimeException)
```

So all my further calls will fail, and the only thing i can do here is to go to server, run `netstat -tulpn` and `kill <node PID>`.

Is there anything you can do about it?

### PR DESCRIPTION
**Pull request by jasonlfunk**
As mentioned in the issue, I am not sure how to write a test for it.

Closes #120

### CODE REVIEW

File: src/ZombieDriver.php [hunk 1]
```diff
@@ -32,6 +32,12 @@ class ZombieDriver extends CoreDriver
      */
     public function __construct($serverOrHost, $port = null)
     {
+        // Ensure that the "stop" method gets called no matter
+        // how the PHP execution finishes.
+        if (function_exists("register_shutdown_function")) {
```
Comment (aik099) [inline src/ZombieDriver.php]: Is `IF` necessary?

File: src/ZombieDriver.php [hunk 2]
```diff
@@ -32,6 +32,12 @@ class ZombieDriver extends CoreDriver
...
+          register_shutdown_function(array($this, 'stop'));
```
Comment (aik099) [inline src/ZombieDriver.php]: Please also check if session is started before attempting to stop it.

Comment (jasonlfunk) [inline src/ZombieDriver.php]: There is no need to. The Driver makes sure that there is a server before stopping it and the Server checks if the server is running before stopping it. I don't think there is a need for another layer of redundancy.

File: src/ZombieDriver.php (hunk 1 above)
Comment (jasonlfunk) [inline src/ZombieDriver.php]: Necessary? I guess not. It looks like it's available in 4 and 5.

**Comment by aik099** [comment]
Looks good to me. Does it solve the issue you're having @jasonlfunk ?

@stof , can you please take a look at this too?

**Comment by jasonlfunk** [comment]
@aik099 It solved my problem.

**Comment by aik099** [comment]
Does it also solve the issue, where there has been a fatal error during test execution?

**Comment by stof** [comment]
Couldn't we use ``__destruct`` instead of keeping a reference to the driver until shutdown (which prevents GC) ?

**Comment by aik099** [comment]
I thought driver is used all the time, so it can't be garbage collected anyway?

I'm not sure, that `__destruct` is really called in case of abnormal script termination, like when Fatal Error happens. I guess this needs to be tested to be certain.

**Comment by stof** [comment]
@aik099 it can if you delete the reference to the Session before the end of the process

**Comment by aik099** [comment]
@jasonlfunk , can you please make @stof 's recommended adjustments and test if that still works?

**Comment by jasonlfunk** [comment]
@aik099 In the event of a fatal error, the `__destruct` callback is not called. A simple call confirmed this:

````
class Test {
        function __construct() {
                register_shutdown_function(array($this, "shutdown"));
        }

        function __destruct() {
                file_put_contents("/tmp/destruct.txt", "yes");
        }

        function shutdown() {
                file_put_contents("/tmp/shutdown.txt", "yes");
        }

        function trigger_fatal() {
                $nothing->fatal();
        }
}

$test = new Test();
$test->trigger_fatal();
````

/tmp/shutdown.txt exists.
/tmp/destruct.txt does not exist.

**Comment by aik099** [comment]
As I suspected. The `__destruct` is only called during garbage collection and application shutdown most likely isn't right time to garbage collect because we can free app memory anyway.

@stof , any idea you we can do https://github.com/minkphp/MinkZombieDriver/pull/125#issuecomment-73667425 and use `register_shutdown_function` at same time?

**Comment by aik099** [comment]
@stof , any update on this? In the PR we're keeping reference to ZombieDriver, not session. So all should be fine I guess.

Anything else is preventing this PR from being merged?

**Comment by robertfausk** [comment]
@stof Will there be any update on this PR?

**Comment by aik099** [comment]
PR looks good to me. At least it doesn't make things worse.

**Comment by aik099** [comment]
Merging, thanks @jasonlfunk .

**Pull request merged**